### PR TITLE
packet injector: add TCP support

### PR DESCRIPTION
**Pull request by masco-enovance**

### CODE REVIEW

**Comment by masco-enovance** [comment]
rerun tests

**Comment by masco-enovance** [comment]
rerun tests

File: statics/js/components/inject-form.js [hunk 1]
```diff
@@ -33,6 +34,16 @@ Vue.component('inject-form', {
         <label for="inject-count">ICMP Identifier</label>\
         <input id="inject-id" type="number" class="form-control input-sm" v-model="id" min="0" />\
       </div>\
+      <div v-if="true">\
+        <div class="form-group">\
+          <label for="src-port">Src Port</label>\
```
Comment (safchain) [inline statics/js/components/inject-form.js]: this should be only displayed if TCP

File: statics/js/components/inject-form.js [hunk 2]
```diff
@@ -33,6 +34,16 @@ Vue.component('inject-form', {
         <label for="inject-count">ICMP Identifier</label>\
```
Comment (safchain) [inline statics/js/components/inject-form.js]: This should be not present if TCP

File: packet_injector/inject_packet.go [hunk 3]
```diff
@@ -134,8 +138,33 @@ func InjectPacket(pp *PacketParams, g *graph.Graph) (string, error) {
...
+	case "tcp":
+		ethLayer.EthernetType = layers.EthernetTypeIPv4
+		ipLayer := &layers.IPv4{SrcIP: srcIP, DstIP: dstIP, Version: 4, Protocol: layers.IPProtocolTCP, TTL: 64}
+		srcPort := layers.TCPPort(pp.SrcPort)
+		dstPort := layers.TCPPort(pp.DstPort)
+		tcpLayer := &layers.TCP{
+			SrcPort: srcPort,
+			DstPort: dstPort,
+			Urgent:  0,
+			Seq:     11050,
```
Comment (safchain) [inline packet_injector/inject_packet.go]: this should be random

File: packet_injector/inject_packet.go [hunk 4]
```diff
@@ -134,8 +138,33 @@ func InjectPacket(pp *PacketParams, g *graph.Graph) (string, error) {
...
 		l = append(l, ethLayer, ipLayer, icmpLayer, payload)
+	case "tcp":
+		ethLayer.EthernetType = layers.EthernetTypeIPv4
+		ipLayer := &layers.IPv4{SrcIP: srcIP, DstIP: dstIP, Version: 4, Protocol: layers.IPProtocolTCP, TTL: 64}
+		srcPort := layers.TCPPort(pp.SrcPort)
+		dstPort := layers.TCPPort(pp.DstPort)
+		tcpLayer := &layers.TCP{
+			SrcPort: srcPort,
+			DstPort: dstPort,
+			Urgent:  0,
```
Comment (safchain) [inline packet_injector/inject_packet.go]: no need to set it to 0 as 0 is the default value for int

File: packet_injector/inject_packet.go [hunk 5]
```diff
@@ -134,8 +138,33 @@ func InjectPacket(pp *PacketParams, g *graph.Graph) (string, error) {
...
+			Seq:     11050,
+			Ack:     0,
```
Comment (safchain) [inline packet_injector/inject_packet.go]: ditto

File: statics/js/components/inject-form.js (hunk 1 above)
Comment (masco-enovance) [inline statics/js/components/inject-form.js]: hi @safchain , i need help on this when i try to hide the div by using the value of select's option i am getting "Unexpected identifier". below is the code at line number 37,
<div v-if="type === 'tcp'">
need some hint on this

File: packet_injector/inject_packet.go (hunk 4 above)
Comment (masco-enovance) [inline packet_injector/inject_packet.go]: done

File: packet_injector/inject_packet.go (hunk 3 above)
Comment (masco-enovance) [inline packet_injector/inject_packet.go]: generating random value using rand.Uint32() method.

File: packet_injector/inject_packet.go (hunk 5 above)
Comment (masco-enovance) [inline packet_injector/inject_packet.go]: removed

File: packet_injector/inject_packet.go [hunk 6]
```diff
@@ -134,8 +138,33 @@ func InjectPacket(pp *PacketParams, g *graph.Graph) (string, error) {
...
+			ACK:     false,
```
Comment (masco-enovance) [inline packet_injector/inject_packet.go]: removed

File: packet_injector/inject_packet.go [hunk 7]
```diff
@@ -134,8 +138,33 @@ func InjectPacket(pp *PacketParams, g *graph.Graph) (string, error) {
...
+			SYN:     false,
```
Comment (masco-enovance) [inline packet_injector/inject_packet.go]: switched to true.

File: statics/js/components/inject-form.js (hunk 1 above)
Comment (safchain) [inline statics/js/components/inject-form.js]: Hi @masco-enovance, I'll have a look at this, will send a PR on your branch

**Comment by safchain** [comment]
You need to update the command line too

File: cmd/client/packet_injector.go [hunk 8]
```diff
@@ -95,7 +99,9 @@ func addInjectPacketFlags(cmd *cobra.Command) {
 	cmd.Flags().StringVarP(&dstIP, "dstIP", "", "", "destination node IP")
 	cmd.Flags().StringVarP(&srcMAC, "srcMAC", "", "", "source node MAC")
 	cmd.Flags().StringVarP(&dstMAC, "dstMAC", "", "", "destination node MAC")
-	cmd.Flags().StringVarP(&packetType, "type", "", "icmp", "packet type: icmp4")
+	cmd.Flags().Int64VarP(&srcPort, "srcPort", "", 0, "source port for TCP packet")
+	cmd.Flags().Int64VarP(&dstPort, "dstPort", "", 0, "destination port for TCP packet")
+	cmd.Flags().StringVarP(&packetType, "type", "", "icmp4", "packet type: icmp4/6, tcp4/6")
```
Comment (safchain) [inline cmd/client/packet_injector.go]: maybe better to give the exact options like:

packet type: icmp4, icmp6, tcp4, tcp6

File: api/packet_injector.go [hunk 9]
```diff
@@ -35,6 +36,11 @@ import (
 	"github.com/skydive-project/skydive/validator"
 )
 
+const (
+	min = 1024
+	max = 15565
```
Comment (safchain) [inline api/packet_injector.go]: why 15565 ? should not be 65535 ?

File: api/packet_injector.go [hunk 10]
```diff
@@ -117,12 +125,24 @@ func (pi *PacketInjectorAPI) requestToParams(ppr *PacketParamsReq) (string, *pac
 		}
 	}
 
+	if ppr.Type == "tcp4" || ppr.Type == "tcp6" {
+		port := rand.Int63n(max-min) + min
+		if ppr.SrcPort == 0 {
+			ppr.SrcPort = port
+		}
+		if ppr.DstPort == 0 {
```
Comment (safchain) [inline api/packet_injector.go]: I would prefer to have src and dst with different random ports

**Pull request merged**